Mathlib/RingTheory/LocalRing/Subring.lean

```diff
@@ -0,0 +1,120 @@
+/-
+Copyright (c) 2024 Andrew Yang, Yaël Dillies, Javier López-Contreras. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang, Yaël Dillies, Javier López-Contreras
+-/
+import Mathlib.Tactic.FieldSimp
+import Mathlib.RingTheory.LocalRing.RingHom.Basic
+import Mathlib.RingTheory.Localization.AtPrime
+
+
+/-!
+# Local subrings of fields
+
+# Main result
+- `LocalSubring` : The class of local subrings of a commutative ring.
+- `LocalSubring.ofPrime`: The localization of a subring as a `LocalSubring`.
+-/
+
+open IsLocalRing Set
+
+open scoped Polynomial
+
+
+variable {R S : Type*} [CommRing R] [CommRing S]
+variable {K : Type*} [Field K]
+
+instance [Nontrivial S] (f : R →+* S) (s : Subring R) [IsLocalRing s] : IsLocalRing (s.map f) :=
+  .of_surjective' (f.restrict s _ (fun _ ↦ Set.mem_image_of_mem f))
+    (fun ⟨_, a, ha, e⟩ ↦ ⟨⟨a, ha⟩, Subtype.ext e⟩)
+
+instance isLocalRing_top [IsLocalRing R] : IsLocalRing (⊤ : Subring R) :=
+  Subring.topEquiv.symm.isLocalRing
+
+variable (R) in
+/-- The class of local subrings of a commutative ring. -/
+@[ext]
+structure LocalSubring where
+  /-- The underlying subring of a local subring. -/
+  toSubring : Subring R
+  [isLocalRing : IsLocalRing toSubring]
+
+namespace LocalSubring
+
+attribute [instance] isLocalRing
+
+lemma toSubring_injective : Function.Injective (toSubring (R := R)) := by
+  rintro ⟨a, b⟩ ⟨c, d⟩ rfl; rfl
+
+/-- Copy of a local subring with a new `carrier` equal to the old one.
+Useful to fix definitional equalities. -/
+protected def copy (S : LocalSubring R) (s : Set R) (hs : s = ↑S.toSubring) : LocalSubring R :=
+  LocalSubring.mk (S.toSubring.copy s hs) (isLocalRing := hs ▸ S.2)
+
+/-- The image of a `LocalSubring` as a `LocalSubring`. -/
+@[simps! toSubring]
+def map [Nontrivial S] (f : R →+* S) (s : LocalSubring R) : LocalSubring S :=
+  mk (s.1.map f)
+
+/-- The range of a ringhom from a local ring as a `LocalSubring`. -/
+@[simps! toSubring]
+def range [IsLocalRing R] [Nontrivial S] (f : R →+* S) : LocalSubring S :=
+  .copy (map f (mk ⊤)) f.range (by ext x; exact congr(x ∈ $(Set.image_univ.symm)))
+
+/--
+The domination order on local subrings.
+`A` dominates `B` if and only if `B ≤ A` and `m_A ∩ B = m_B`.
+-/
+@[stacks 00I9]
+instance : PartialOrder (LocalSubring R) where
+  le A B := ∃ h : A.1 ≤ B.1, IsLocalHom (Subring.inclusion h)
```

Comment:
Could you please add a lemma `LocalSubring.le_iff` that allows to rewrite `A \le B` into this definition, and please put a similar docstring on that lemma as on this instance.